module/zfs/zfs_ioctl.c

```diff
@@ -3512,6 +3512,68 @@ zfs_ioc_log_history(const char *unused, nvlist_t *innvl, nvlist_t *outnvl)
 	return (error);
 }
 
+/*
+ * This ioctl is used to set the bootenv configuration on the current
+ * pool. This configuration is stored in the second padding area of the label,
+ * and it is used by the FreeBSD-derived bootloader used on Illumos when
+ * determining what dataset to boot. The bootenv configuratoin has 4 main
+ * elements. The first is the command to pass to the bootloader. This is
+ * primarily to be used to tell the bootloader what dataset to use as the
+ * bootfs. If this isn't used, the bootloader will use the bootfs property
+ * from the pool.  The second is the environment map. We use this to pass a
+ * few environment variables to the bootloader.  The third component is the
+ * number of boots to attempt before switching to the bootenv configuration.
+ * The final component is the number of boots that have been attempted.
+ *
+ * When we boot, we first increment the number of attempted boots. If the
+ * maximum number of boots is 0, or the number of attempted boots is equal to
+ * the maximum boot count, we clear the bootenv configuration and use the
+ * command and environment map to control the boot. Otherwise, we proceed with
+ * boot as normal.
+ */
+/* ARGSUSED */
+static const zfs_ioc_key_t zfs_keys_set_bootenv[] = {
+	{"envmap",	DATA_TYPE_STRING,	0},
+};
+
+static int
+zfs_ioc_set_bootenv(const char *name, nvlist_t *innvl, nvlist_t *outnvl)
+{
+	char *envmap;
+	int error;
+	spa_t *spa;
+
+	if (nvlist_lookup_string(innvl, "envmap", &envmap) != 0)
+		return (EINVAL);
```

Comment:
You can actually safely use `fnvlist_lookup_string` here.  Since the "envmap" entry in the `zfs_ioc_key_t zfs_keys_set_bootenv` was not marked `ZK_OPTIONAL` all of the keys are considered mandatory and their  types will have been verified by this point.